augur/filter/include_exclude_rules.py

```diff
@@ -188,24 +189,34 @@ def filter_by_query(metadata: pd.DataFrame, query: str) -> FilterFunctionReturn:
     # Create a copy to prevent modification of the original DataFrame.
     metadata_copy = metadata.copy()
 
-    # Try extracting column names from the query.
-    variables = extract_variables(query)
-    if variables is not None:
-        columns = variables.intersection(metadata_copy.columns)
+    # Convert data types before applying the query.
+    if column_types:
+        # If types are explicitly provided, use those as source of truth.
+        for column, dtype in column_types.items():
+            if dtype == 'int' or dtype == 'float':
+                metadata_copy[column] = pd.to_numeric(metadata_copy[column], errors='ignore')
+            elif dtype == 'str':
+                metadata_copy[column] = metadata_copy[column].astype(str)
     else:
-        # Use all columns.
-        columns = metadata_copy.columns
-
-    # Support numeric comparisons in query strings.
-    #
-    # The built-in data type inference when loading the DataFrame does not
-    # support nullable numeric columns, so numeric comparisons won't work on
-    # those columns. pd.to_numeric does proper conversion on those columns, and
-    # will not make any changes to columns with other values.
-    #
-    # TODO: Try boolean conversion?
-    for column in columns:
-        metadata_copy[column] = pd.to_numeric(metadata_copy[column], errors='ignore')
+        # Otherwise, try converting all columns to numeric.
+        # This should cover most use cases, since one common problem is that the
+        # built-in data type inference when loading the DataFrame does not
+        # support nullable numeric columns, so numeric comparisons won't work on
+        # those columns. pd.to_numeric does proper conversion on those columns,
+        # and will not make any changes to columns with other values.
+        #
+        # TODO: Try boolean conversion?
+
+        # Try extracting column names from the query.
+        variables = extract_variables(query)
+        if variables is not None:
+            columns = variables.intersection(metadata_copy.columns)
+        else:
+            # FIXME: raise custom error class and handle separately in apply_filters
+            raise AugurError("Unable to determine what columns are used in the query.")
+
+        for column in columns:
+            metadata_copy[column] = pd.to_numeric(metadata_copy[column], errors='ignore')
```

Comment:
My initial expectation was that `--query-columns` would be used to _augment_ the fallback behaviour of "convert all used columns to numeric", not replace it wholesale. Does that make sense? Did you consider that behaviour instead?

I'm also somewhat surprised that we'd raise an error here instead of keeping the ultimate fallback of converting all columns (i.e. since we don't know what is used) to numeric. 

To sum up,  Basically something like:

```python
if column_types is None:
    column_types = {}

variables = extract_variables(query)
if variables is not None:
    columns = variables.intersection(metadata_copy.columns)
else:
    columns = metadata_copy.columns
    # or alternatively: columns = [] to make an exhaustive --query-columns
    # mandatory only when extract_variables() fails

for column in columns:
    column_types.setdefault(column, 'float') # or 'numeric' if taking my other suggestion
```

before the application of `column_types` conversion done above.

Reply:
I didn't consider this, but it seems better. I'll try it out.

Reply:
Updated in [force-push](https://github.com/nextstrain/augur/compare/64a54743a12f744ee873523f4ec40d1986e3ad88..679519ee5092df34f7658b4e677bb8c5b87f69d2). I chose to not make an exhaustive --query-columns mandatory when extract_variables() fails. This keeps the existing behavior of automatic type inference.

https://github.com/nextstrain/augur/blob/679519ee5092df34f7658b4e677bb8c5b87f69d2/augur/filter/include_exclude_rules.py#L196-L211